### PR TITLE
Detect system's CPU/Memory to adjust TiKV's config automatically

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Adjust block-cache-size by system's total memory, adjust end-point-pool/sched-worker-pool size by system's total cpu cores.
ref https://github.com/pingcap/tikv/issues/1597
@siddontang @hhkbp2 PTAL

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -403,8 +410,16 @@ fn get_rocksdb_write_cf_option(config: &toml::Value) -> RocksdbOptions {
...
 
-fn get_rocksdb_raftlog_cf_option(config: &toml::Value) -> RocksdbOptions {
-    get_rocksdb_cf_option(config, "raftcf", 256 * 1024 * 1024, false, false)
+fn get_rocksdb_raftlog_cf_option(config: &toml::Value, total_mem: u64) -> RocksdbOptions {
+    let mut default_block_cache_size = (total_mem as f64 * 0.01) as u64;
+    if default_block_cache_size < RAFTCF_MIN_MEM {
+        default_block_cache_size = RAFTCF_MIN_MEM;
+    }
+    if default_block_cache_size > RAFTCF_MAX_MEM {
+        default_block_cache_size = RAFTCF_MAX_MEM;
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: max 2GB is proper here? if a user want to use more, we should also allow it.

**Comment by siddontang** [comment]
how about write memory buffer?

File: src/bin/tikv-server.rs (hunk 1 above)
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: `default_block_cache_size` takes effect in the condition that `block-cache-size` is not set in config file.

**Comment by zhangjinpeng1987** [comment]
The default write buffer size is suitable for most cases, no need adjust by system's total memory.

**Comment by siddontang** [comment]
Em, for aws t2.micro in our stability test, the total memory is 1GB, so here the total memory buffer will exceed the max memory value.

**Comment by overvenus** [comment]
Cc #1183

**Comment by ngaut** [comment]
ping

**Comment by zhangjinpeng1987** [comment]
This pr aim to reduce the repeated config works for most cases, not for all cases, tiny machine is not include.

**Comment by hhkbp2** [comment]
LGTM

**Comment by siddontang** [comment]
Please fix the conflict.

PTAL @overvenus

File: Cargo.toml [hunk 2]
```diff
@@ -56,6 +56,7 @@ clap = "2"
 url = "1.2"
 getopts = "0.2"
 regex = "0.1"
+sysinfo = "0.2.5"
```
Comment (overvenus) [inline Cargo.toml]: Why not `0.2.7`?

https://crates.io/crates/sysinfo

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ extern crate signal;
 #[cfg(unix)]
 extern crate nix;
 extern crate prometheus;
+extern crate sysinfo;
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: It only supports Linux and macOS, how about other platforms?
https://github.com/GuillaumeGomez/sysinfo#system-

File: src/storage/config.rs [hunk 4]
```diff
@@ -14,7 +14,7 @@
 const DEFAULT_STORE_PATH: &'static str = "";
 const DEFAULT_SCHED_CAPACITY: usize = 10240;
 const DEFAULT_SCHED_MSG_PER_TICK: usize = 1024;
-const DEFAULT_SCHED_CONCURRENCY: usize = 102400;
+const DEFAULT_SCHED_CONCURRENCY: usize = 1024000;
```
Comment (overvenus) [inline src/storage/config.rs]: 10x boost?

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -789,7 +829,12 @@ fn main() {
     let pd_client = RpcClient::new(&pd_endpoints).unwrap();
     let cluster_id = pd_client.cluster_id;
 
-    let mut cfg = build_cfg(&matches, &config, cluster_id, addr);
+    let mut system = System::new();
+    system.refresh_all();
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: No need to `refresh_all` again.

https://github.com/GuillaumeGomez/sysinfo/blob/37829411d4186264c24dfa96f15522655d5060bb/src/linux/system.rs#L70

File: src/bin/tikv-server.rs [hunk 6]
```diff
@@ -383,17 +388,19 @@ fn get_rocksdb_cf_option(config: &toml::Value,
     opts
 }
 
-fn get_rocksdb_default_cf_option(config: &toml::Value) -> RocksdbOptions {
+fn get_rocksdb_default_cf_option(config: &toml::Value, total_mem: u64) -> RocksdbOptions {
     // Default column family uses bloom filter.
+    let default_block_cache_size = (total_mem as f64 * 0.4) as u64;
```
Comment (overvenus) [inline src/bin/tikv-server.rs]: Should we define those memory ratios as a const f64 array and make sure they never exceed 1?

File: src/storage/config.rs (hunk 4 above)
Comment (ngaut) [inline src/storage/config.rs]: Is this necessary? I guess 102400 is enough.

File: src/bin/tikv-server.rs (hunk 3 above)
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: any suggestion?

File: src/bin/tikv-server.rs (hunk 5 above)
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Need to call `refresh_all` for 0.2.5 .

File: src/bin/tikv-server.rs (hunk 6 above)
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Good suggestion.

File: src/bin/tikv-server.rs (hunk 3 above)
Comment (siddontang) [inline src/bin/tikv-server.rs]: refer `util/clocktime.rs`

**Comment by zhangjinpeng1987** [comment]
sysinfo has bug. 
ref https://github.com/GuillaumeGomez/sysinfo/issues/50

File: src/bin/tikv-server.rs [hunk 7]
```diff
@@ -28,7 +28,7 @@ extern crate signal;
 #[cfg(unix)]
 extern crate nix;
 extern crate prometheus;
-extern crate sysinfo;
+extern crate sys_info;
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: does sys_info support multi platforms? like unix, mac, and windows?

Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Yes, it supports Linux, Mac OS X and Windows.

File: src/bin/tikv-server.rs [hunk 8]
```diff
@@ -60,6 +62,22 @@ use tikv::raftstore::store::keys::region_raft_prefix_len;
 use tikv::pd::RpcClient;
 use tikv::util::time_monitor::TimeMonitor;
 
+const RAFTCF_MIN_MEM: u64 = 256 * 1024 * 1024;
+const RAFTCF_MAX_MEM: u64 = 2 * 1024 * 1024 * 1024;
+const KB: u64 = 1024;
+const DEFAULT_BLOCK_CACHE_RATIO: &'static [f64] = &[0.4, 0.15, 0.01];
```
Comment (andelf) [inline src/bin/tikv-server.rs]: should use bit shift instead of float operation,
for alignment with MB / GB boundary.

Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: It is ratio

Comment (andelf) [inline src/bin/tikv-server.rs]: what if 8173720 * 0.01? this results in an ugly number in Byte.

Comment (andelf) [inline src/bin/tikv-server.rs]: For a normal TiDB user, setting this value to 81737 is ugly and weird.

```
0.4  r shifts 4 or 5
0.15 r shifts 6
0.01 r shifts 6 or 7
```

File: src/bin/tikv-server.rs [hunk 9]
```diff
@@ -60,6 +62,27 @@ use tikv::raftstore::store::keys::region_raft_prefix_len;
...
+        ratio = ratio + v;
+    }
+    if ratio > 1.0 {
+        return false;
+    }
+    true
+}
+
+fn align_to_mb(n: u64) -> u64 {
+    let m = n >> 20;
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: can we use (n & 0xFFFFFFFFFFF00000) directly?

**Pull request merged**